### DISCUSSION TITLE
[Bug]: The connection only works once after a restart

### DISCUSSION THREAD

**Discussion by psi33** [wlcrs/huawei_solar#888] [Connecting to the inverter]
### Describe the issue

First of all, many thanks for this ingenious integration. Unfortunately, it only works once for me and is not stable. After restarting the integration, all data is correctly retrieved from the inverter via the network. 

If I try to connect again, the values are no longer updated. This is clearly visible in the log.

A second home server also accesses the same interface via the network. This shows that the connection to the Modbus interface is working.

For a test, I ended this second query so that the Home Assistant has unrestricted sole access. Unfortunately, this did not result in any improvement and the values are only read in once after a restart. Why does the integration crash? 

Thank you for your help

### Describe your Huawei Solar Setup

Inverter Type: SUN2000-10KTL-M1
Inverter Firmware version: V100R001C00SPC162
sDongle present: Yes
sDongle Type: sDongleA-05 (WiFi / Ethernet)
sDongle Connectivitiy: Ethernet
sDongle Firmware: V100R001C00SPC335
Power meter present: three phase
Optimizers Present: No
Battery:  None
Huawei Solar integration version: None

### How do you connect to the inverter?

Via the SDongle, wired connection

### Upload your Diagnostics File

Drag & Drop your Diagnostics File here.
[config_entry-huawei_solar-01JFN1QA6R2EV2RNH153ECHFY9.json](https://github.com/user-attachments/files/18246641/config_entry-huawei_solar-01JFN1QA6R2EV2RNH153ECHFY9.json)
[Log.txt](https://github.com/user-attachments/files/18246644/Log.txt)

### Upload your relevant debug logs

```text
[Log.txt](https://github.com/user-attachments/files/18246644/Log.txt)
```

### Please confirm the following:

- [X] I'm running the latest release of Home Assistant.
- [X] I'm running the latest release of this integration.
- [X] I did not find an existing issue describing this problem.
- [X] I did upload the diagnostics-file that I could retrieve from  the 'Devices & Services Page'
- [X] I did upload the relevant debug logs (via 'Enable Debug Logging'-feature or by manually configuring HA logging)

**Comment by NBOnGithub** [comment]
I'm seeing the same behaviour for the past week. Any restart of home assistant gives me a single snapshot value for the battery status and inferior power, etc. and nothing further. Command actions still seem to take effect as I can manage to change the charge from grid settings via automations. 
Previous to this week the integration has been working flawlessly. 
I've tried resetting the solar system by turning off the power to it but this has not fixed it either

**Comment by Mav70** [comment]
@psi33 I'm also seeing the same issue for the past couple of days. Any update on a resolution?

**Comment by NBOnGithub** [comment]
I eventually removed the integration and also removed the download from hacs. Then I rebooted and reinstalled and it started working again. 
It might be getting something corrupt on disk that it is then failing to reload properly.
Note that I did need to do a second reboot to fix the automation warnings.

**Comment by psi33** [comment]
Thx @NBOnGithub 
I have deleted the integration and also deleted it in HACS. Then restarted HA. Added the integration (1.5.0a1) again. It has improved and works for about 10 queries, but even after that the integration crashes. Does anyone have an idea where this could be

**Comment by Mav70** [comment]
@psi33 @NBOnGithub I've also tried to delete and reinstall but with the same results as before. What I have noticed though is that sometimes the inverter isn't responding to pings (I connect via wifi) although I can access it fine through the app.  The wifi connection at the inverter side seems to go into a "sleep" mode.

**Comment by psi33** [comment]
@NBOnGithub und @Mav70 
I connect via a fixed network cable. I have just tested it again. After restarting the integration (1.5.0a1), I get 20 measuring points from the inverter. Then the integration crashes without a log entry and does not send any further data. In all this time I have not lost a single ping. So I think we can rule out the inverter and the network connection. What else can we test or adjust? Does anyone have an idea?

**Comment by psi33** [comment]
Unfortunately, the new HA version 2025.1 does not bring any improvement. I think only the developer can help us here. I hope for feedback

**Comment by fehlerteufel** [comment]
I have a similar issue (worked fine before updating HA and the integration yesterday). Data from the inverter is updated normally, but data from the additionally connected power meter is only updated once after restarting the integration.

**Comment by WILHER** [comment]
I have the same Problem. After reboot ist works für a couple of seconds than it crash

**Comment by fehlerteufel** [comment]
Reverting the Huawei Solar Integration back to version 1.3.3 solved the problem for now. I'm happy to help in resolving the problem with version 1.4.1, please ping me if I can support in some way.

**Comment by psi33** [comment]
Unfortunately, installing version 1.3.3 does not help me. The data is updated for 5 minutes and then the interface crashes.
I also hope for a solution soon and will gladly support you in testing and reporting.

**Comment by psi33** [comment]
@wlcrs 
What does this mean for me if the entry was migrated from Issue to discussion?

**Reply by psi33** [reply]
Hi @wlcrs 
First and foremost, I would like to sincerely thank you for your outstanding work on this integration. It is of great value to us, and we truly appreciate your effort and dedication.

Some time ago, you converted my issue into a discussion, for which I am also grateful. However, we have not received any further updates regarding the next steps since then.

It would be very helpful for us to know if the reported issue might be due to a bug in the interface or if there is a workaround we could apply. I would greatly appreciate it if you could provide us with a brief update on how to proceed.

If additional time and effort are required to achieve a working solution, I want to let you know that I am willing to support this financially as well.

Thank you very much in advance for your time and assistance. I look forward to hearing from you.

Best regards,
Peter

**Comment by psi33** [comment]
@wlcrs 
I have installed 1.5.3 and unfortunately the integration is still not working. How can I support you, I would like to help?

@WILHER and @fehlerteufel 
Have you also installed it, how does it look for you?

**Reply by WILHER** [reply]
I also installed 1.5.3, unfortunately the same problem.
The connection is established after a restart for approx. 1 minute and goes into "not available"

**Comment by psi33** [comment]
I immediately installed 2025.2.0 today, unfortunately the connection is also started here and after a minute “not available” appears. It's a pity that we never hear a word from the developer @wlcrs about the next steps and whether we can provide support. :-(

**Reply by amaduain** [reply]
> I immediately installed 2025.2.0 today, unfortunately the connection is also started here and after a minute “not available” appears. It's a pity that we never hear a word from the developer @wlcrs about the next steps and whether we can provide support. :-(

Sadly this is also my case, @psi33 sometimes the open source world can be frustrating, but @wlcrs has been working to fix this modbus version compatibility hell, so let's give him some kudos and then help him to continue moving forward.

As I said, I'm in the same place, when I try to install the new module 1.5.3 over HA 2025.0 it says it can't connect.
In my case I'm using a Protoss PW11 TCP/Modbus adapter, and I can see the connection on the admin web page, but the counters are not going up, so it seems that the integration is not seeing that is really connected:
![image](https://github.com/user-attachments/assets/865d6067-7787-4f62-8af5-7573b16c9da9)

So I tried to use the python library that also was maintained by @wlcrs : https://pypi.org/project/huawei-solar/

After creating and environment, the following code is working:

`pip3 -m asyncio
>>> import asyncio
>>> from huawei_solar import AsyncHuaweiSolar, register_names as rn
>>> slave_id = 1
>>> client = await AsyncHuaweiSolar.create("192.168.8.123", 502, slave_id)
>>> result = await client.get(rn.NB_PV_STRINGS, slave_id)
>>> result
Result(value=2, unit=None)
>>>
`
So that is working, from the same host that HA is installed, and I can see that the TCP adapter is receiving and querying the modbus correctly:
![image](https://github.com/user-attachments/assets/5c83ff8a-2728-496f-9fc3-4c14ca46f4eb)

Also note that the other connection from the HA to connect via Huawei module is still there.
Anyway, if you need more feedback or testing let me know, but it seems that cover this use case will take a while.

**Comment by N1ckn4m3sN3t** [comment]
Same Problem for me - then i changed connection to SUN2000-xxx instead of SDongle and seems that's working with described settings:

"When connecting to the inverter AP the host IP is typically 192.168.200.1 and the slave id is typically 0"

**Reply by persching1978** [reply]
I use the connection of my SUN2000-xxx and not the SDongle connection, but it's not working with huawei solar V1.5.3 and HA2025.2.1. 
So there seems to be another difference that you can successfully connect and others can't. I use currently huawei solar V1.4.1 and this is working stable for me.

**Comment by N1ckn4m3sN3t** [comment]
After the initial problems, I tested the fork https://github.com/tmenguy/huawei_solar, but it didn't work with that either. Then I went back to the original integration and since then (about 2h^^) it works great.

 SUN2000-8KTL-M101074300-002
Firmware: V100R001C00SPC153

Home Assistant
    Core 2025.2.1
    Supervisor 2025.02.0
    Operating System 14.2
    Frontend 20250205.0

Huawei solar 1.5.3

**Reply by persching1978** [reply]
I have a SUN2000-10KTL-M1 with Firmware V100R001C00SPC167 and the MBUS Firmware V100R001C00SPC335. Maybe there is the difference if it's working with 1.5.3. But before HA2025.1.0 I used Huawei Solar V1.5.3, but I have to try it often to start the integration. Once started it works if I do not restart HA...

**Comment by psi33** [comment]
@wlcrs 
I have installed HA 2025.2.3 and Huawei 1.5.4. The connection works and the data is read in correctly. But after about a minute, unavailable appears. My setting is shown in the first post. 

How and with what information can I provide support?
Is there an alternative to Huawei Solar if this inverter is not supported?

I am grateful for any advice.

If this integration runs stably for a month, I will also be happy to donate a sum of money

**Comment by psi33** [comment]
Hello everyone 
It is a pity that we have not received any feedback from the developer even after 2 months. Does anyone know an alternative solution for a Modbus query? How do you deal with the problem?

**Reply by NBOnGithub** [reply]
Since I reinstalled the add on completely in December I've had one or two drop outs but they've resolved after reboots. The issue has not reoccured for me but I am terrified it will every time I update HA.
The only thing I can suggest is you have another go at completely removing everything to do with the integration and then reinstall it again 🤞 

SUN2000-5KTL-L1
Firmware: V200R001C00SPC148

Home Assistant
Core 2025.2.4
Supervisor 2025.02.1
Operating System 14.2
Frontend 20250214.0

Huawei solar 1.5.4

**Reply by amaduain** [reply]
> Hello everyone It is a pity that we have not received any feedback from the developer even after 2 months. Does anyone know an alternative solution for a Modbus query? How do you deal with the problem?

Well, there is a library  for python, that is the foundation of the HACS component (developed by the same team) that you can use to troubleshoot and test, I built some kind of script with much less functionality that put the info I needed on the MQTT broker.

The library is this one: https://github.com/wlcrs/huawei-solar-lib?tab=readme-ov-file
You can build something that suits your needs if this kind of issue repeats (it will repeat, because the devs needs to stay updated with the HA Core python changes IMO)

You can find a sample code here: https://gist.github.com/amaduain/acbbd78603c07e614eb4796e2e42de40

**Comment by psi33** [comment]
Thanks for all the feedback. I have just studied the Modbus documentation from Huawei. Commissioned the Modbus register manually with the integrated function. The whole thing has been running stably for a day now. I will continue to monitor it.

**Reply by Kinyuu** [reply]
Can you provide details on how you did it? I'm having the exact same issue...

**Reply by psi33** [reply]
@Kinyuu 

Copy this code into your Configuration.yaml and adjust the name and IP address before. This has been working very reliably for me for quite some time.

modbus:
 - name: Your Name
    type: tcp
    host: your IP
    port: 502
    delay: 5 
    timeout: 10 
    sensors:
      - name: WR_PV_Active_Power
        address: 32080
        input_type: holding
        data_type: int32
        unit_of_measurement: W
        state_class: measurement
        scale: 1
      - name: WR_PV_Daily_Energy_Yield
        address: 32114
        input_type: holding
        data_type: uint32
        unit_of_measurement: kWh
        state_class: total_increasing
        scale: 0.01
      - name: WR_PV_Accumulated_Energy_Yield
        address: 32106
        input_type: holding
        data_type: uint32
        unit_of_measurement: kWh
        state_class: total_increasing
        scale: 0.01
      - name: WR_PV_Peak_Active_Power_Current_Day
        address: 32078
        input_type: holding
        data_type: int32
        unit_of_measurement: W
        state_class: measurement
        scale: 1
      - name: WR_PV_1_Volt
        address: 32016
        input_type: holding
        data_type: int16
        unit_of_measurement: V
        state_class: measurement
        scale: 0.1
      - name: WR_PV_1_Ampere
        address: 32017
        input_type: holding
        data_type: int16
        unit_of_measurement: A
        state_class: measurement
        scale: 0.1
      - name: WR_PV_2_Volt
        address: 32018
        input_type: holding
        data_type: int16
        unit_of_measurement: V
        state_class: measurement
        scale: 0.1
      - name: WR_PV_2_Ampere
        address: 32019
        input_type: holding
        data_type: int16
        unit_of_measurement: A
        state_class: measurement
        scale: 0.1
      - name: WR_PV_Phase_A_Ampere
        address: 32072
        input_type: holding
        data_type: int32
        unit_of_measurement: A
        state_class: measurement
        scale: 0.001
      - name: WR_PV_Phase_B_Ampere
        address: 32074
        input_type: holding
        data_type: int32
        unit_of_measurement: A
        state_class: measurement
        scale: 0.001
      - name: WR_PV_Phase_C_Ampere
        address: 32076
        input_type: holding
        data_type: int32
        unit_of_measurement: A
        state_class: measurement
        scale: 0.001
      - name: WR_PV_Internal_Temperature
        address: 32087
        input_type: holding
        data_type: int16
        unit_of_measurement: °C
        state_class: measurement
        scale: 0.1
      - name: WR_PV_Positive_Active_Power
        address: 37119
        input_type: holding
        data_type: int32
        unit_of_measurement: kW
        state_class: measurement
        scale: 0.01
      - name: WR_PV_Reverse_Active_Power
        address: 37121
        input_type: holding
        data_type: int32
        unit_of_measurement: kW
        state_class: measurement
        scale: 0.01

**Reply by Kinyuu** [reply]
@psi33 
But that is modbus through IP, my problem is through RS485, through IP the integration works flawlessly

**Comment by Kinyuu** [comment]
I have exactly the same issues, the first or first few times it reads, but then it never reads again until I reboot.
Checking the logs, it seems the inverter timeouts the connection, and then it never connects again.
I tested the connection manually with my PC connected to the inverter, reading on address 40000, and it reads fine, but if I "spam" the read button, it timeouts some of the commands, I'm wondering if the library is flooding the inverter with reads every 30s and somehow this is timing out the connection or something...
In the meanwhile, switched it to over network and it works fine, but I would prefer the serial connection...

As a side note, I connected the power meter DDSU666 to RS485A1 and B1, instead of the normal A2 and B2 on the SUN2000-5KTL, and, wouldn't you know it, IT happened with the power meter! The meter would report consumption to the inverter a few times, and then after a minute or so, the inverter would generate an alarm that the power meter was not reachable! Switched it to A2 and B2, and it worked flawlessly again.

I have no idea how there are people with successful RS485 serial modbus connections working.
